Added #[projections(primitive)] attribute for structures. (#14699)
coq/coq: doc/sphinx/language/core/records.rst

   :cmd:`Record` and :cmd:`Structure` are synonyms.

   This command supports the :attr:`universes(polymorphic)`,
   :attr:`universes(template)`, :attr:`universes(cumulative)`,

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Author: Huh, I had no idea! :-)

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer: You got this backwards, @jfehrle: the Oxford comma is typical in AmE and somewhat less common in BrE; see https://en.wikipedia.org/wiki/Serial_comma

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Author: OK now I am confused. I actually agree with @jfehrle on the comma convention here, since it is what I am used too. However, before I changed anything in this document, the attributes were listed with the extra `,`. Which one should be ought to use?

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer 2: @cpitclaudel Wikipedia is mistaken.  The Oxford comma is not commonly used in US English--not in journalism, not in general-interest books and not in the few academic papers I've read.  Nor have I have seen it much in the Coq documentation.

@Zimmi48 Any thoughts?

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer 3: > Wikipedia is mistaken.

I didn't check the references, but I'm not expecting this claim to be wrong:

>A majority of American style guides mandate use of the serial comma. These include: APA style,[8] The Chicago Manual of Style, Garner's Modern American Usage,[9] The MLA Style Manual, Strunk and White's The Elements of Style,[10] and the U.S. Government Printing Office Style Manual. By contrast, the Associated Press Stylebook and The New York Times Style Book[11] advise against it.

Also, for the record, by own grammar checking tool, which is set to American English also recommends it.

That being said, since this is apparently a debated topic, I would tend to side with Jim against the Oxford comma. But I would also not lose much time over this.

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer: @jfehrle Well, I'm no expert, so I usually go by published sources; here's what my copy of "Garner's Modern English Usage" has to say (sorry, I don't have an electronic copy):

![image](https://user-images.githubusercontent.com/8181630/127564792-f5ad7da3-be63-4beb-8ea1-29660d91e301.png)

![image](https://user-images.githubusercontent.com/8181630/127564963-c013700f-779d-413b-8538-cc7d0a3d85f9.png)

FWIW, Coq is copyrighted to `The Coq development team, INRIA, CNRS, and contributors`

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer 2: @cpitclaudel I'm familiar with the discussion.  It's just that the Oxford comma is rarely used in practice.  As Wikipedia notes, the US style guides are not unanimous.  Actual usage trumps any "authoritative" source.

The part about "When the members are compound, calling for _and_ within themselves, clarity demands the final comma." would be the main exception in my experience.

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer: Eh, we have actual usage right in front of us: the way the manual is currently written.  So I see no need to change it to conform to perceptions about what's more or less commonly used out there.

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Author: I had a look back through the git blame and it appears that @Zimmi48 wrote it first in the style @jfehrle suggests, it seems later @ejgallego changed it to have the `,` before the `and`.

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer 2: @cpitclaudel I would not select the documentation as an exemplar of excellent, well-edited writing from which you can infer what is common usage.

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer: Indeed not, and hence there's no need to get into these long arguments about comma usage preferences :) Sorry that I got dragged into this — I was originally reacting only to the comment that this would "generally be considered incorrect", which isn't true, AFAICT.

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer 2: Indeed, "incorrect" was a bit too emphatic.

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer 3: I suggest settling for inconsistency on this topic, that is, let the PR author do what they want, and avoid large-scale changes relative to these commas.

>    :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.

Reviewer 3: > I had a look back through the git blame and it appears that @Zimmi48 wrote it first in the style @jfehrle suggests, it seems later @ejgallego changed it to have the `,` before the `and`.

That is: there is no Oxford comma in all copyright statements (manual and header files) but they appear in the author fields of the `.opam` files. Not really a "change" IMHO.

   Note here that the fields ``Rat_bottom_cond`` depends on the field ``bottom``
   and ``Rat_irred_cond`` depends on both ``top`` and ``bottom``.

> Let us now see the work done by the :cmd:`Record` macro. First the macro

Author: I'll check for them.

> Let us now see the work done by the :cmd:`Record` macro. First the macro

Author: Alright, replaced two more occurances.

Primitive Projections
~~~~~~~~~~~~~~~~~~~~~

.. flag:: Primitive Projections

   This :term:`flag` turns on the use of primitive

>    projections when defining subsequent records (even through the ``Inductive``

Reviewer 3: Maybe keeping the mention that this applies even with the records defined with the `Inductive` or `CoInductive` commands would be useful (actually, it is kept in the description of the `flag`...). By the way, you forgot to list the attribute in the supported attributes for the `Inductive` command.

Author: I didn't mention it for Inductive since it is technically not supported. I believe only for the `:= { .. }` syntax.

Author: Alright, I've modified all the other mentions of the attribute to specify that it is valid only when using record syntax. This includes `Inductive`.

> construct is disabled (see :ref:`primitive_projections` though).

Reviewer 3: The comment in the parentheses comes from the previous text, but now it becomes self-referential and useless.

Author: Alright, removed.

> To eliminate the (co-)inductive type, one must use its defined primitive 

Reviewer 3: Wouldn't it read better like this?
```suggestion
To eliminate the record type, one must use its defined primitive 
```

projections.

> .. The following paragraph is quite redundant with what is above

Reviewer 3: Do you still consider this comment to be relevant? It seems to me that it could be removed after your changes.

Author: I will keep it as I think it's useful when skimming the documentation.

For compatibility, the parameters still appear to the user when printing terms
even though they are absent in the actual AST manipulated by the kernel. This
can be changed by unsetting the :flag:`Printing Primitive Projection Parameters`
flag.

There are currently two ways to introduce primitive records types:

#. Through the :cmd:`Record` command, in which case the type has to be
   non-recursive. The defined type enjoys eta-conversion definitionally,
   that is the generalized form of surjective pairing for records:
   `r` ``= Build_``\ `R` ``(``\ `r`\ ``.(``\ |p_1|\ ``) …`` `r`\ ``.(``\ |p_n|\ ``))``.
   Eta-conversion allows to define dependent elimination for these types as well.
#. Through the :cmd:`Inductive` and :cmd:`CoInductive` commands, when
   the :term:`body` of the definition is a record declaration of the form
   ``Build_``\ `R` ``{`` |p_1| ``:`` |t_1|\ ``; … ;`` |p_n| ``:`` |t_n| ``}``.
   In this case the types can be recursive and eta-conversion is disallowed.
   These kind of record types differ from their traditional versions in the
   sense that dependent
   elimination is not available for them and only non-dependent case analysis

>    can be defined.

Author: This seems to be the case:
```coq

#[projections(primitive)]
Inductive foo := {
 bar : Type;
}.
```
yields
```
foo : Type

foo is not universe polymorphic
foo has primitive projections without eta conversion.
Expands to: Inductive Top.foo
```
I believe the `Inductive` command disallows eta conversion to take place.

   primitive projection `p` applied to a record object `r` (i.e., primitive
   projections are always applied). Even if the record type has parameters,
   these do not appear in the internal representation of applications of the
   projection, considerably reducing the sizes of terms when manipulating
   parameterized records and type checking time. On the user level, primitive
   projections can be used as a replacement for the usual defined ones, although
   there are a few notable differences.

.. attr:: projections(primitive{? = {| yes | no } })
   :name: projections(primitive)

   This :term:`boolean attribute` can be used to override the value of the
   :flag:`Primitive Projections` :term:`flag` for the record type being
   defined.

.. flag:: Printing Primitive Projection Parameters

   This compatibility :term:`flag` reconstructs internally omitted parameters at
   printing time (even though they are absent in the actual AST manipulated
   by the kernel).

Reduction
+++++++++

The basic reduction rule of a primitive projection is
|p_i| ``(Build_``\ `R` |t_1| … |t_n|\ ``)`` :math:`{\rightarrow_{\iota}}` |t_i|.

> However, to take the δ flag into account, projections can be in two states:
> folded or unfolded. An unfolded primitive projection application obeys the rule
> above, while the folded version delta-reduces to the unfolded version. This
> allows to precisely mimic the usual unfolding rules of :term:`constants <constant>`. Projections obey the usual ``simpl`` flags of the :cmd:`Arguments`
> command in particular.

Reviewer 2: ```suggestion
However, to take the δ flag into account, projections can be in two states:
folded or unfolded. An unfolded primitive projection application obeys the rule
above, while the folded version delta-reduces to the unfolded version. This
allows to precisely mimic the usual unfolding rules of :term:`constants <constant>`.
Projections obey the usual ``simpl`` flags of the :cmd:`Arguments`
command in particular.
```

long line.  (The wording is not great, but let's not touch that today.)

#. Through the ``Inductive`` and ``CoInductive`` commands, when
   the :term:`body` of the definition is a record declaration of the form
   ``Build_``\ `R` ``{`` |p_1| ``:`` |t_1|\ ``; … ;`` |p_n| ``:`` |t_n| ``}``.
   In this case the types can be recursive and eta-conversion is disallowed.
   These kind of record types differ from their traditional versions in the
   sense that dependent
   elimination is not available for them and only non-dependent case analysis
   can be defined.

For both cases the :flag:`Primitive Projections` :term:`flag` must be set or
the :attr:`projections(primitive)` :term:`attribute`  must be supplied.

.. flag:: Primitive Projections

   This :term:`flag` turns on the use of primitive projections when defining

>    subsequent records (even through the ``Inductive`` and ``CoInductive``

Reviewer 3: ```suggestion
   subsequent records (even through the :cmd:`Inductive` and :cmd:`CoInductive`
```

Author: There were quite a few of these littered around the document, I've changed as many as I could find.

   elimination is not available for them and only non-dependent case analysis
   can be defined.

> For both cases the :term:`flag` :flag:`Primitive Projections` must be set or 
> the :term:`attribute` :attr:`projections(primitive)` must be supplied.

Reviewer 3: ```suggestion
For both cases the :flag:`Primitive Projections` :term:`flag` must be set or 
the :attr:`projections(primitive)` :term:`attribute` must be supplied.
```

Compatibility Projections and :g:`match`
++++++++++++++++++++++++++++++++++++++++

To ease compatibility with ordinary record types, each primitive projection is
also defined as an ordinary :term:`constant` taking parameters and an object of
the record type as arguments, and whose :term:`body` is an application of the
unfolded primitive projection of the same name. These constants are used when
elaborating partial applications of the projection. One can distinguish them
from applications of the primitive projection if the :flag:`Printing Primitive
Projection Parameters` flag is off: For a primitive projection application,
parameters are printed as underscores while for the compatibility projections
they are printed as usual.

Additionally, user-written :g:`match` constructs on primitive records are
desugared into substitution of the projections, they cannot be printed back as

> :g:`match` constructs.

Reviewer 2: This seems to be only a reformatting without a user-visible wording change.  While it's reasonable to split really long lines, reformatting in this case makes for more work in the review (i.e., figuring out whether there is a wording change) and makes it harder to identify the commits that made substantive changes.  Better not to change this now.

Author: I have not changed the wording here, only the formatting. I should have perhaps done the formatting in a later commit and squashed after review.

Author: The reason I have done these, is because I was splitting up a large line. In my IDE (VSCode) I just press a shortcut and it hard wraps the entire paragraph. I'll try to keep the diff clearer next time.

Reviewer 2: > I have not changed the wording here, only the formatting. I should have perhaps done the formatting in a later commit and squashed after review.

Slightly better, but I would not reformat even in a separate commit absent a good reason to do so.  (And the principle applies even more strongly to source code.)
